DOC-10138 v23.1.21 release notes (#18518)
cockroachdb/docs: src/current/_includes/releases/v23.1/v23.1.21.md

## v23.1.21

Release Date: May 7, 2024

{% include releases/new-release-downloads-docker-image.md release=include.release %}

<h3 id="v23-1-21-sql-language-changes">SQL language changes</h3>

- The `FORCE_INVERTED_INDEX` hint causes the [optimizer]({% link v23.1/cost-based-optimizer.md %}) to prefer a query plan scan over any [inverted index]({% link v23.1/inverted-indexes.md %}) of the hinted table. An error is emitted if no such query plan can be generated. [#122301][#122301]

> - Introduced three new [cluster settings]({% link v24.1/cluster-settings.md %}) for controlling [table statistics]({% link v23.1/cost-based-optimizer.md %}#table-statistics) forecasting:

Reviewer: ```suggestion
- Introduced three new [cluster settings]({% link v23.1/cluster-settings.md %}) for controlling [table statistics]({% link v23.1/cost-based-optimizer.md %}#table-statistics) forecasting:
```

    - [`sql.stats.forecasts.min_observations`]({% link v23.1/cluster-settings.md %}#setting-sql-stats-forecasts-min-observations) is the minimum number of observed statistics required to produce a forecast.

>     - [`sql.stats.forecasts.min_goodness_of_fit`]({% link v23.1/cluster-settings.md %}#setting-sql-stats-forecasts-min-goodness-of-fit) is the minimum R² (goodness of fit) measurement required from all predictive models to use a forecast.

Reviewer: Does the superscript render correctly? If not, I would use `R<sup>2</sup>`.

    - [`sql.stats.forecasts.max_decrease`]({% link v23.1/cluster-settings.md %}#setting-sql-stats-forecasts-max-decrease) is the most a prediction can decrease, expressed as the minimum ratio of the prediction to the lowest prior observation. [#122990][#122990]

> - Added a [session variable]({% link v23.1/set-vars.md %}) `optimizer_use_improved_multi_column_selectivity_estimate`, which if enabled, causes the [optimizer]({% link v23.1/cost-based-optimizer.md %}) to use an improved selectivity estimate for multi-column predicates. This setting will default to `true` on versions 24.2+, and `false` on prior versions. [#123068][#123068]

Reviewer: ```suggestion
- Added a [session variable]({% link v23.1/set-vars.md %}) `optimizer_use_improved_multi_column_selectivity_estimate`, which if enabled, causes the [optimizer]({% link v23.1/cost-based-optimizer.md %}) to use an improved selectivity estimate for multi-column predicates. This setting will default to `true` on v24.2 and later, and `false` on prior versions. [#123068][#123068]
```

<h3 id="v23-1-21-operational-changes">Operational changes</h3>

> - A minimum [Raft scheduler]({% link v23.1/architecture/replication-layer.md %}#raft) concurrency is now enforced per store so that nodes with many stores do not spread workers too thin. This helps to avoid high scheduler latency across replicas on a store when load is imbalanced. [#120797][#120797]

Reviewer: This is a hyper-nit but I think it's more grammatically correct to use an adverb

```suggestion
- A minimum [Raft scheduler]({% link v23.1/architecture/replication-layer.md %}#raft) concurrency is now enforced per store so that nodes with many stores do not spread workers too thinly. This helps to avoid high scheduler latency across replicas on a store when load is imbalanced. [#120797][#120797]
```

Reviewer: I'd maybe also link "store" to https://www.cockroachlabs.com/docs/v23.2/architecture/storage-layer#overview as we have done elsewhere.

<h3 id="v23-1-21-bug-fixes">Bug fixes</h3>

- Fixed a bug introduced in v22.2.9 that could cause a slow memory leak that can accumulate when opening many new connections. [#121056][#121056]

> - [Sequence]({% link v23.1/create-sequence.md %}) options for `NO MINVALUE` and `NO MAXVALUE` now match PostgreSQL behavior. Sequence `MINVALUE` and `MAXVALUE` automatically adjust to new types bounds mirroring behavior of PostgreSQL. [#121307][#121307]

Reviewer: It's not clear what `NO MINVALUE` and `NO MAXVALUE` do. I'm also unclear what "new types bounds" means.

```suggestion
- [Sequence]({% link v23.1/create-sequence.md %}) options for `NO MINVALUE` and `NO MAXVALUE` now match PostgreSQL behavior. Sequence `MINVALUE` and `MAXVALUE` automatically adjust to new types bounds, matching PostgreSQL behavior. [#121307][#121307]
```

Author: updated to clarify "new types bounds". I looked at the [tests in the code](https://github.com/cockroachdb/cockroach/pull/121307/files#diff-9c15b618f9f47149f21bc53c432b9357a273c2983485bbdfc66d98def43d7a34R353) and manually ran the tests on v23.1.20:
```
florence@defaultdb> CREATE SEQUENCE seq_change_type_bounds MINVALUE -9223372036854775808 MAXVALUE 9223372036854775807 START 0;
CREATE SEQUENCE

Time: 37ms total (execution 14ms / network 23ms)

florence@defaultdb> SHOW CREATE seq_change_type_bounds;
        table_name       |                                                       create_statement
-------------------------+-------------------------------------------------------------------------------------------------------------------------------
  seq_change_type_bounds | CREATE SEQUENCE public.seq_change_type_bounds MINVALUE -9223372036854775808 MAXVALUE 9223372036854775807 INCREMENT 1 START 0
(1 row)

Time: 58ms total (execution 34ms / network 24ms)

florence@florence-aws-east2-mwb.aws-us-east-2.cockroachlabs.cloud:26257/defaultdb> ALTER SEQUENCE seq_change_type_bounds AS smallint;
ERROR: MINVALUE (-9223372036854775808) must be greater than (-32768) for type INT2
```

In v24.1.0-beta-3 which also has this fix, the `ALTER SEQUENCE` does not get the error. It runs successfully and automatically adjusts the bounds of the new integer datatype, in this case, `smallint`.

- Fixed a bug where the [timeseries graphs shown on the **SQL Activity Statement Fingerprint** page]({% link v23.1/ui-statements-page.md %}#charts) in the [DB Console]({% link v23.1/ui-overview.md %}) were not rendering properly. This involved fixing a bug related to setting the time range of the charts. [#121382][#121382] [#122235][#122235]

> - CockroachDB could previously incorrectly evaluate `IN` expressions that had `INT2` or `INT4` type on the left side and values outside of the range of the left side on the right side. The bug has been present since at least v21.1 and is now fixed. [#121955][#121955]

Reviewer: ```suggestion
- Fixed a bug where CockroachDB could incorrectly evaluate `IN` expressions that had `INT2` or `INT4` type on the left side, and values on the right side that were outside the range of the left side. The bug had been present since at least v21.1. [#121955][#121955]
```

Author: updated, much clearer.

- Previously, on long-running [sessions]({% link v23.1/show-sessions.md %}) that issue many (hundreds of thousands or more) [transactions]({% link v23.1/transactions.md %}), CockroachDB's internal memory accounting system, the limit for which is configured via the [`--max-sql-memory` flag]({% link v23.1/cockroach-start.md %}#general) could leak. This bug, in turn, could result in the error message `"root: memory budget exceeded"` for other queries. The bug was present in v23.1.17 and is now fixed. [#121949][#121949] [#122235][#122235]
- Reintroduced [cluster setting]({% link v23.1/cluster-settings.md %}) `sql.auth.modify_cluster_setting_applies_to_all.enabled` so that mixed-version clusters can migrate off of this setting, which is deprecated in favor of the privilege [`MODIFYSQLCLUSTERSETTING`]({% link v23.1/set-cluster-setting.md %}#required-privileges). [#122055][#122055] [#122635][#122635]
- Fixed a bug where a [`GRANT ... ON ALL TABLES`]({% link v23.1/grant.md %}) statement could fail if sequences existed and they did not support a privilege (e.g., `BACKUP`). [#122057][#122057]

> - Fixed a bug where [client certificate authentication]({% link v23.1/authentication.md %}#client-authentication) combined with [identity maps]({% link v24.1/sso-sql.md %}#identity-map-configuration) (`server.identity_map.configuration`) did not work since v23.1. For the feature to work correctly, the client must specify a valid db user in the [connection string]({% link v23.1/connection-parameters.md %}). [#122746][#122746]

Reviewer: ```suggestion
- Fixed a bug where [client certificate authentication]({% link v23.1/authentication.md %}#client-authentication) combined with [identity maps]({% link v24.1/sso-sql.md %}#identity-map-configuration) (`server.identity_map.configuration`) did not work. For the feature to work correctly, the client must specify a valid database user in the [connection string]({% link v23.1/connection-parameters.md %}). This bug had been present since v23.1.[#122746][#122746]
```

- Added a new [session variable]({% link v23.1/set-vars.md %}) `optimizer_use_improved_zigzag_join_costing`. When enabled, the cost of [zigzag joins]({% link v23.1/cost-based-optimizer.md %}#zigzag-joins) is updated so zigzag joins will only be chosen over scans if the zigzag joins produce fewer rows. This change only applies if the session variable `enable_zigzag_join` is also `on`. [#123068][#123068]
- Improved the selectivity estimation of multi-column filters by the [optimizer]({% link v23.1/cost-based-optimizer.md %}) when the multi-column distinct count is high. This avoids cases where CockroachDB significantly over-estimates the selectivity of a multi-column predicate and as a result can prevent the optimizer from choosing a bad query plan. [#123068][#123068]

<h3 id="v23-1-21-contributors">Contributors</h3>

This release includes 59 merged PRs by 26 authors.
We would like to thank the following contributors from the CockroachDB community:

> - Andrew Delph (first-time contributor)

Reviewer: I think we decided to stop naming the contributors. Let me know if you heard differently.

Author: removed  the named contributor.
